File: backend/iliad/models/enums.py
"""Controlled vocabularies used across the EHR.

These are plain ``str`` enums so they serialize as human-readable strings in the
API and are easy for agents to reason about. Keep values lowercase-snake and
stable — agents will filter on them.
"""

from __future__ import annotations

from enum import Enum


class EncounterStatus(str, Enum):
    planned = "planned"
    in_progress = "in-progress"  # active admission
    finished = "finished"
    cancelled = "cancelled"


class EncounterClass(str, Enum):
    ambulatory = "AMB"
    inpatient = "IMP"
    emergency = "EMER"
    observation = "OBSENC"
    virtual = "VR"


class DispositionType(str, Enum):
    """Where a patient is expected/decided to go after discharge."""

    home = "home"
    home_with_services = "home_with_services"  # home + home health / DME
    snf = "snf"  # skilled nursing facility
    assisted_living = "assisted_living"
    inpatient_rehab = "inpatient_rehab"  # IRF / acute rehab
    ltac = "ltac"  # long-term acute care
    hospice_home = "hospice_home"
    hospice_facility = "hospice_facility"
    ama = "ama"  # against medical advice
    expired = "expired"
    undetermined = "undetermined"


class DispositionStatus(str, Enum):
    """Lifecycle of the discharge-planning process for an encounter."""

    undetermined = "undetermined"  # no plan yet
    predicted = "predicted"  # agent/team has a leading prediction
    decided = "decided"  # team committed to a disposition
    in_progress = "in_progress"  # actively arranging (calls, auths, orders)
    ready = "ready"  # all barriers cleared, ready to discharge
    discharged = "discharged"


class ConditionCategory(str, Enum):
    problem_list_item = "problem-list-item"
    encounter_diagnosis = "encounter-diagnosis"


class ClinicalStatus(str, Enum):
    active = "active"
    recurrence = "recurrence"
    relapse = "relapse"
    inactive = "inactive"
    remission = "remission"
    resolved = "resolved"


class ObservationCategory(str, Enum):
    vital_signs = "vital-signs"
    laboratory = "laboratory"
    imaging = "imaging"
    survey = "survey"


class ObservationStatus(str, Enum):
    registered = "registered"
    preliminary = "preliminary"
    pending = "pending"  # ordered, specimen/result not yet available
    final = "final"
    amended = "amended"
    cancelled = "cancelled"


class ReportStatus(str, Enum):
    registered = "registered"
    partial = "partial"
    preliminary = "preliminary"
    pending = "pending"
    final = "final"
    cancelled = "cancelled"


class MedicationStatus(str, Enum):
    active = "active"
    on_hold = "on-hold"
    completed = "completed"
    stopped = "stopped"
    draft = "draft"  # ordered but not yet signed/active
    cancelled = "cancelled"


class NoteType(str, Enum):
    progress = "progress"
    history_and_physical = "history_and_physical"
    discharge_summary = "discharge_summary"
    consult = "consult"
    case_management = "case_management"
    nursing = "nursing"
    social_work = "social_work"
    therapy = "therapy"  # PT/OT/SLP eval
    after_visit_summary = "after_visit_summary"
    family_communication = "family_communication"  # calls/discussions with family re: dispo preferences


class NoteStatus(str, Enum):
    draft = "draft"  # pended, not signed
    signed = "signed"
    amended = "amended"


class OrderType(str, Enum):
    lab = "lab"
    medication = "medication"
    imaging = "imaging"
    consult = "consult"  # e.g. PM&R, PT/OT, social work
    nursing = "nursing"  # e.g. isolation, foley care
    dispo = "dispo"  # discharge / placement order
    referral = "referral"


class OrderStatus(str, Enum):
    """Order lifecycle.

    ``draft`` == "pended" in Epic parlance: saved but not signed. Agents create
    orders as drafts for a clinician to sign, or sign directly when authorized.
    """

    draft = "draft"  # pended, awaiting signature
    signed = "signed"  # signed, active/in-progress (e.g. lab collected)
    completed = "completed"  # fulfilled / resulted
    cancelled = "cancelled"
    discontinued = "discontinued"


class TaskType(str, Enum):
    call_snf = "call_snf"
    call_family = "call_family"
    call_pcp = "call_pcp"
    order_lab = "order_lab"
    draft_consult = "draft_consult"
    insurance_auth = "insurance_auth"
    collect_preference = "collect_preference"
    verify_eligibility = "verify_eligibility"
    arrange_transport = "arrange_transport"
    other = "other"


class TaskStatus(str, Enum):
    pending = "pending"
    in_progress = "in_progress"
    blocked = "blocked"
    completed = "completed"
    cancelled = "cancelled"


class FacilityType(str, Enum):
    snf = "snf"
    assisted_living = "assisted_living"
    inpatient_rehab = "inpatient_rehab"
    ltac = "ltac"
    hospice = "hospice"
    home_health = "home_health"
    dme = "dme"  # durable medical equipment


class CommunicationDirection(str, Enum):
    outbound = "outbound"
    inbound = "inbound"


class CommunicationModality(str, Enum):
    phone = "phone"
    sms = "sms"
    fax = "fax"
    email = "email"
    portal = "portal"


class PartyType(str, Enum):
    family = "family"
    patient = "patient"
    snf = "snf"
    facility = "facility"
    pcp = "pcp"
    insurance = "insurance"
    other = "other"

File: frontend/src/lib/enums.ts
// Controlled vocabularies mirrored from backend/ehr/models/enums.py.
// Keep values in sync with the API — they are sent verbatim as query params / body fields.

export interface EnumOption {
  value: string;
  label: string;
}

function opts(pairs: [string, string][]): EnumOption[] {
  return pairs.map(([value, label]) => ({ value, label }));
}

export const ENCOUNTER_STATUS = opts([
  ["planned", "Planned"],
  ["in-progress", "In progress"],
  ["finished", "Finished"],
  ["cancelled", "Cancelled"],
]);

export const ENCOUNTER_CLASS = opts([
  ["AMB", "Ambulatory"],
  ["IMP", "Inpatient"],
  ["EMER", "Emergency"],
  ["OBSENC", "Observation"],
  ["VR", "Virtual"],
]);

export const DISPOSITION_TYPE = opts([
  ["home", "Home"],
  ["home_with_services", "Home + services"],
  ["snf", "Skilled nursing facility"],
  ["assisted_living", "Assisted living"],
  ["inpatient_rehab", "Inpatient rehab (IRF)"],
  ["ltac", "Long-term acute care"],
  ["hospice_home", "Hospice — home"],
  ["hospice_facility", "Hospice — facility"],
  ["ama", "Against medical advice"],
  ["expired", "Expired"],
  ["undetermined", "Undetermined"],
]);

export const DISPOSITION_STATUS = opts([
  ["undetermined", "Undetermined"],
  ["predicted", "Predicted"],
  ["decided", "Decided"],
  ["in_progress", "In progress"],
  ["ready", "Ready"],
  ["discharged", "Discharged"],
]);

export const CONDITION_CATEGORY = opts([
  ["problem-list-item", "Problem list item"],
  ["encounter-diagnosis", "Encounter diagnosis"],
]);

export const CLINICAL_STATUS = opts([
  ["active", "Active"],
  ["recurrence", "Recurrence"],
  ["relapse", "Relapse"],
  ["inactive", "Inactive"],
  ["remission", "Remission"],
  ["resolved", "Resolved"],
]);

export const OBSERVATION_CATEGORY = opts([
  ["vital-signs", "Vital signs"],
  ["laboratory", "Laboratory"],
  ["imaging", "Imaging"],
  ["survey", "Survey"],
]);

export const OBSERVATION_STATUS = opts([
  ["registered", "Registered"],
  ["preliminary", "Preliminary"],
  ["pending", "Pending"],
  ["final", "Final"],
  ["amended", "Amended"],
  ["cancelled", "Cancelled"],
]);

export const MEDICATION_STATUS = opts([
  ["active", "Active"],
  ["on-hold", "On hold"],
  ["completed", "Completed"],
  ["stopped", "Stopped"],
  ["draft", "Draft"],
  ["cancelled", "Cancelled"],
]);

export const NOTE_TYPE = opts([
  ["progress", "Progress"],
  ["history_and_physical", "History & physical"],
  ["discharge_summary", "Discharge summary"],
  ["consult", "Consult"],
  ["case_management", "Case management"],
  ["nursing", "Nursing"],
  ["social_work", "Social work"],
  ["therapy", "Therapy (PT/OT/SLP)"],
  ["after_visit_summary", "After-visit summary"],
]);

export const NOTE_STATUS = opts([
  ["draft", "Draft"],
  ["signed", "Signed"],
  ["amended", "Amended"],
]);

export const ORDER_TYPE = opts([
  ["lab", "Lab"],
  ["medication", "Medication"],
  ["imaging", "Imaging"],
  ["consult", "Consult"],
  ["nursing", "Nursing"],
  ["dispo", "Disposition"],
  ["referral", "Referral"],
]);

export const ORDER_STATUS = opts([
  ["draft", "Draft (pended)"],
  ["signed", "Signed"],
  ["completed", "Completed"],
  ["cancelled", "Cancelled"],
  ["discontinued", "Discontinued"],
]);

export const ORDER_PRIORITY = opts([
  ["routine", "Routine"],
  ["urgent", "Urgent"],
  ["stat", "STAT"],
]);

export const TASK_TYPE = opts([
  ["call_snf", "Call SNF"],
  ["call_family", "Call family"],
  ["call_pcp", "Call PCP"],
  ["order_lab", "Order lab"],
  ["draft_consult", "Draft consult"],
  ["insurance_auth", "Insurance authorization"],
  ["collect_preference", "Collect preference"],
  ["verify_eligibility", "Verify eligibility"],
  ["arrange_transport", "Arrange transport"],
  ["other", "Other"],
]);

export const TASK_STATUS = opts([
  ["pending", "Pending"],
  ["in_progress", "In progress"],
  ["blocked", "Blocked"],
  ["completed", "Completed"],
  ["cancelled", "Cancelled"],
]);

export const TASK_PRIORITY = opts([
  ["low", "Low"],
  ["medium", "Medium"],
  ["high", "High"],
]);

export const FACILITY_TYPE = opts([
  ["snf", "Skilled nursing facility"],
  ["assisted_living", "Assisted living"],
  ["inpatient_rehab", "Inpatient rehab"],
  ["ltac", "Long-term acute care"],
  ["hospice", "Hospice"],
  ["home_health", "Home health"],
  ["dme", "Durable medical equipment"],
]);

export const COMM_DIRECTION = opts([
  ["outbound", "Outbound"],
  ["inbound", "Inbound"],
]);

export const COMM_MODALITY = opts([
  ["phone", "Phone"],
  ["sms", "SMS"],
  ["fax", "Fax"],
  ["email", "Email"],
  ["portal", "Portal"],
]);

export const PARTY_TYPE = opts([
  ["family", "Family"],
  ["patient", "Patient"],
  ["snf", "SNF"],
  ["facility", "Facility"],
  ["pcp", "PCP"],
  ["insurance", "Insurance"],
  ["other", "Other"],
]);

function toLabelMap(list: EnumOption[]): Record<string, string> {
  return Object.fromEntries(list.map((o) => [o.value, o.label]));
}

export const LABELS = {
  encounterStatus: toLabelMap(ENCOUNTER_STATUS),
  encounterClass: toLabelMap(ENCOUNTER_CLASS),
  dispositionType: toLabelMap(DISPOSITION_TYPE),
  dispositionStatus: toLabelMap(DISPOSITION_STATUS),
  conditionCategory: toLabelMap(CONDITION_CATEGORY),
  clinicalStatus: toLabelMap(CLINICAL_STATUS),
  observationCategory: toLabelMap(OBSERVATION_CATEGORY),
  observationStatus: toLabelMap(OBSERVATION_STATUS),
  medicationStatus: toLabelMap(MEDICATION_STATUS),
  noteType: toLabelMap(NOTE_TYPE),
  noteStatus: toLabelMap(NOTE_STATUS),
  orderType: toLabelMap(ORDER_TYPE),
  orderStatus: toLabelMap(ORDER_STATUS),
  taskType: toLabelMap(TASK_TYPE),
  taskStatus: toLabelMap(TASK_STATUS),
  facilityType: toLabelMap(FACILITY_TYPE),
  commModality: toLabelMap(COMM_MODALITY),
  partyType: toLabelMap(PARTY_TYPE),
};

export function humanize(value?: string | null): string {
  if (!value) return "";
  return value
    .replace(/[_-]/g, " ")
    .replace(/\b\w/g, (c) => c.toUpperCase());
}
